src/main/java/org/mitre/synthea/export/BB2RIFExporter.java

```diff
@@ -0,0 +1,4667 @@
+     * @param codeToMap the Synthea code to look for
+     * @param rand a source of random numbers used to pick one of the list of BFD codes
+     * @return the BFD code or null if the code can't be mapped
+     */
+    public String map(String codeToMap, RandomNumberGenerator rand) {
+      return map(codeToMap, "code", rand);
+    }
+
+    /**
+     * Get one of the BFD codes for the supplied Synthea code. Equivalent to
+     * {@code map(codeToMap, "code", rand)}.
+     * @param codeToMap the Synthea code to look for
+     * @param rand a source of random numbers used to pick one of the list of BFD codes
+     * @param stripDots whether to remove dots in codes (e.g. J39.45 -> J3945)
+     * @return the BFD code or null if the code can't be mapped
+     */
+    public String map(String codeToMap, RandomNumberGenerator rand, boolean stripDots) {
+      return map(codeToMap, "code", rand, stripDots);
+    }
+
+    /**
+     * Get one of the BFD codes for the supplied Synthea code.
+     * @param codeToMap the Synthea code to look for
+     * @param bfdCodeType the type of BFD code to map to
+     * @param rand a source of random numbers used to pick one of the list of BFD codes
+     * @return the BFD code or null if the code can't be mapped
+     */
+    public String map(String codeToMap, String bfdCodeType, RandomNumberGenerator rand) {
+      return map(codeToMap, bfdCodeType, rand, false);
+    }
+
+    /**
+     * Get one of the BFD codes for the supplied Synthea code.
+     * @param codeToMap the Synthea code to look for
+     * @param bfdCodeType the type of BFD code to map to
+     * @param rand a source of random numbers used to pick one of the list of BFD codes
+     * @param stripDots whether to remove dots in codes (e.g. J39.45 -> J3945)
+     * @return the BFD code or null if the code can't be mapped
+     */
+    public String map(String codeToMap, String bfdCodeType, RandomNumberGenerator rand,
+            boolean stripDots) {
+      if (!canMap(codeToMap)) {
+        return null;
+      }
+      List<Map<String, String>> options = map.get(codeToMap);
+      int choice = rand.randInt(options.size());
+      String code = options.get(choice).get(bfdCodeType);
+      if (stripDots) {
+        return code.replaceAll("\\.", "");
+      } else {
+        return code;
+      }
+    }
+  }
+
+  /**
+   * Utility class for writing to BB2 writers.
+   */
+  private static class SynchronizedBBLineWriter {
+    
+    private String bbFieldSeparator = "|";
+    private Path path;
+    
+    /**
+     * Construct a new instance.
+     * @param file the writer to write to
+     * @throws IOException if something goes wrong
+     */
+    public SynchronizedBBLineWriter(Path path) {
+      this.path = path;
+    }
+
+    /**
+     * Construct a new instance.
+     * @param file the writer to write to
+     * @throws IOException if something goes wrong
+     */
+    public SynchronizedBBLineWriter(Path path, String separator) {
```

Comment:
Sorry. Not sure why I put the comment on the class. I guess I couldn't find a better place for it.

What I was looking at are the various export methods, like `exportOutpatient`, `exportInpatient`, etc. There are 5 cases where the code syncs on a `SynchronizedBBLineWriter`, sets up a `claimLine` counter, iterates through something and then has an `if` case when `claimLine == 1`. It seems like there is a pattern there that may be worth extracting.

Reply:
We discussed this but we are going to pass. While the current code is a little repetitious, each instance is different enough that generalizing it would be difficult. In addition, the code is still under active development and we aren't sure that this pattern will continue to hold.